### ISSUE TITLE
Visual glitches and unreliable ensureIndexVisible with autoHeight=true

### ISSUE THREAD

**Issue by rfricz** [ag-grid/ag-grid#6860]
Hi AG Grid team,
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => see 'Providing a Reproducible Scenario'
[] feature request => do not use Github for feature requests, see 'Customers of AG Grid'
[] support request => see 'Requesting Community Support'
```

**Providing a Reproducible Scenario**
* https://plnkr.co/edit/AdMg2Xhxy3bjqbf3?preview

**Current behavior**

There are two related issues:
* Visual glitches when switching between grids contained in tabs, done by clicking the tab handles
  * Grids are needlessly re-rendered after a tab switch, it can be seen that the rows have the wrong height at first and only display correctly after a perceivable delay
  * Scroll position is not retained when switching to another tab and then switching back
* The `ensureIndexVisible` method doesn’t do its job reliably, use the "Go to" buttons at the top to see
  * When the destination tab is already active (i.e. no tab switching is involved), the method jumps to the wrong row when called for the first time and only scrolls to the end when called again
  * Unsuprisingly, the same problem occurs when a programmatic tab switch is done and the method is called right after the destination tab is shown (this is just to illustrate the use case)

**Expected behavior**

* Ideally, tab switching shouldn’t trigger a redraw. An alternative is an API to manually suppress what happens in both the just-hidden and the just-shown grid.
* The `ensureIndexVisible` method should ensure visibility regardless of displayed content. An async API would be fine if synchronicity is an issue.

**Please tell us about your environment:**

* **AG Grid version:** 30.0.3

* **Browser:** Chromium 114, Firefox 115

* **Language:** ES6/5

**Comment by StephenCooper** [comment]
Hi @rfricz,

This is a known limitation with autoHeight which we note in our Documentation due to the lazy calculation of row heights.

https://ag-grid.com/javascript-data-grid/row-height/#lazy-height-calculation

> If scrolling up and showing rows for the first time (e.g. the user jumps to the bottom scroll position and then starts slowly scrolling up), then the row positions will jump as the rows coming into view at the top will get resized and the new height will impact the position of all rows beneath it. For example if the row gets resized to be 10 pixels taller, rows below it will get pushed down by 10 rows. If scrolling down this isn't observed as rows below are not in view.
>
> The above are results of Lazy Height Calculation. It is not possible to avoid these effects.

I will leave this open for now though as I understand there is a valid use case here that currently doesn't have a clear workaround or support.

**Comment by rfricz** [comment]
Thanks for looking into this. I've read the row height documentation before and understood that scrolling up and down (i.e. scrolling by arrow or PgUp/Dn keys, mouse wheel or by dragging the scroll thumb) will jump. The `debounceVerticalScrollbar` option helps with this and I see no problem with scrolling.

However, there's no scrolling involved in the Plunker scenario I provided, it really is only about grid changing display state and the "go to" function – the docs don't imply any limitations of these, they only say that scrolling up will jump, but that doesn't bother me.

**Comment by zoheil-ag** [comment]
Hi rfricz,

The easiest way to resolve this is to suppress row virtualization. This means the grid will size all rows and will know how far down to scroll exactly to bring a specific row into view. However this only works for client-side row model and even then is only applicable when you have a small number of rows shown in the grid as you'd like to avoid a larger initial rendering time.

So unless you've suppressed row virtualization (and loaded all data from the server), the grid doesn't know how tall the rows are until it scrolls them into view and sizes them, at which point the row you scrolled into view (or scrolled to by calling ensureIndexVisible API) may be pushed out of view because the rows above it have larger height.

See the example below 
https://plnkr.co/edit/2XNiZ4eNx5wT5gNk?open=main.js

The delay between the calls to ensureIndexVisible allows the auto-height computation logic to run for the visible rows to set their height. This approach then keeps calling the ensureIndexVisible method as many times as necessary until the scrolling settles, which means all rows have sized themselves and no more scrolling is necessary to bring the desired row into view.

Please adjust the delay between the calls to ensureIndexVisible based on the slowest machines you need to support. You'll need to increase the delay to support slower machines which would take longer to render the rows after each call to the ensureIndexVisible API method.

I hope this helps.

Kind regards,

**Comment by zoheil-ag** [comment]
Hi @rfricz,

Thank you for submitting this on Github.

As this has been open without any further interaction for a while, I will be closing this issue.

If the issue has already been resolved by other means, you can ignore this. If however, you are still in need of support you can choose to raise the issue again for the AG Grid community to support.

Kind regards,
Zoheil

**Comment by liborbus** [comment]
Hi @zoheil-ag,

thank you for your support! We have been following up regarding this ticket with the AG Grid Enterprise support team. Both problems related to grids with variable row height reported by @rfricz are still valid and are real showstoppers for us. I'm not sure if it is the case for closing this issue as resolved.

Let me describe our usecase if anybody from the AG Grid community faces the problems and needs your support as well.

Our application uses a grid component to display structured documents having a tree hierarchy of sections, similar to Word. The application displays document paragraphs in a table similar to Excel - each row can have different height and columns display additional information. Grids for open documents are displayed in a tab container similar to Excel worksheets.

Users can quickly navigate to a specific grid row by clicking on the table of contents displayed in a separate pane. Users can also switch between documents using a tabs at the bottom or a shortcut. 

The application keeps the whole data model for each document at the client side. However, we need we need to use AG Grid row virtualization feature to keep acceptable performance even for complex projects containing tens of documents, each having up to tens of thousands of rows. 

Unfortunately, your suggestion does not solve our problems:

1) **Switch between grids**: It takes a while when a user switches between documents by clicking on a document tab because the target grid needs to be completely redrawn. During that time user can see how the height of row changes. Moreover, scroll position is not kept therefore user does not see the same view when he switches back to a document. This is very distracting and leads to unacceptable UX. 

We need to keep grid content when it is not visible to enable fast display after switching between grids. A possible solution could be to add a new option to disable removal of rendered grid rows from the DOM when grid visibility changes (see [dgrid](https://dgrid.io/)), or to provide an API for suppressing and resuming redrawing (see `suspendRender` / `resumeRender` in [Handsontable](https://handsontable.com/)).

2) **Navigate to a row**: When users clicks on a section in the table of contents or presses END then the target row is not visible. Your proposal to scroll iteratively using timeouts is not reliable and would not work on slow machines. This can be seen when setting CPU to “6x slowdown” in the Performance tab of the Chrome Dev Tool. We’ve found in our POC that increasing this timeout above 200ms results in really poor performance and UX when scrolling to the last row in very large (real) documents.

Both issues are combined when users need to switch to a selected row in another grid. 

We need to achieve similar behavior for switching between grids and going to a selected row (e.g. the last row) users experience in Excel.

We need an AG Grid API which reliably focuses and selects a row in a grid with variable row heights.

**Comment by rfricz** [comment]
Here's a [video showing ](https://github.com/ag-grid/ag-grid/assets/36163339/a33aa238-b4b1-49ef-9634-090a49f02ec5) that the timeout based approach to scrolling is not reliable. Only rendering state driven approach would be reliable.
Knowing when the grid is done rendering would also help with a related issue #6895.

**Comment by zoheil-ag** [comment]
Hi @rfricz and @liborbus,

Firstly, thank you for your patience and for continuing to provide context and samples for us. This is highly appreciated.

In regards to my comment above, the workaround provided will evidently not work for all cases (as stated). For slower machines, a longer timeout duration will be required. To mitigate the visible effect of that the overlay was provided in the plunker example. We are aware this approach will mean a slight unnecessary waiting time for faster machines, but this compromise seems warranted (especially utilizing the overlay once again), considering full functionality will be available on slow and fast machines this way. It's "only" a workaround and not an actual feature or full solution.

In any case, we agree that this area of our grid can be improved and we do already have a feature request for this. I have added the label to this issue now. You can follow it on our pipeline page [here](https://www.ag-grid.com/ag-grid-pipeline/).

Please note, in the spirit of transparency, that this is a large feature that requires significant changes to the way the auto-height sizing and grid scrolling works and at this time we can't commit to a specific timeframe.

So in summary, thank you again for raising valid points regarding this and be rest assured that is already on our radar.

Kind regards,
Zoheil

**Comment by liborbus** [comment]
Hi @zoheil-ag,
Thank you for checking this issue internally! We are looking forward to having these problems resolved in the AG Grid. If you need any help let us know.
Regards,
Libor
